### ISSUE TITLE
FE DE FF E7 is interpeted as halt_baddata() on ARM instead of built-in trap for linux-like platforms.

### ISSUE THREAD

**Issue by GameCubeGBA** [NationalSecurityAgency/ghidra#3723]
**Describe the bug**
FE DE FF E7 is a breakpoint command for debuggers. Indeed, most release programs will not try to execute this, but they do exist, usually behind some opaque predicates. However, trying to fix this is difficult, especially when they in fact CAN be reached. This makes analysis of many arm64 binaries difficult.

**To Reproduce**
1. Open an ARM64 armv8 LE binary with functions in them, one of which with the aforementioned byte sequence
2. Try to disassemble the functions
3. Notice how disassembly stops when it hits the breakpoint

**Expected behavior**
Disassembler notes it as a built-in trap or breakpoint

**Screenshots**

**Attachments**
If applicable, please attach any files that caused problems or log files generated by the software.

**Environment (please complete the following information):**
 - OS: macOS 11.6.1
 - Java Version: 11.0.13
 - Ghidra Version: 10.0.4
 - Ghidra Origin: official gre distro
**Additional context**

**Comment by d-millar** [comment]
@GameCubeGBA  At first glance, this seems to me like the correct behavior.  Bytes after a breakpoint (whether implemented as a HLT, BRK, UND, or INT) cannot generally be assumed to be valid opcodes as there is no presumption of flow across the breakpoint.  Admittedly, I am more familiar with x86 than arm64, but, in x86 code, breakpoint equivalents are regularly used to delineate code boundaries.  Disassembling through them would produce erroneous results.  Can you provide a specific example where there is intended flow across a breakpoint?

**Comment by GameCubeGBA** [comment]
<img width="957" alt="Screen Shot 2021-12-07 at 12 10 56 PM" src="https://user-images.githubusercontent.com/95232096/145074878-60f536d1-2b29-49a6-9970-f800f6098c6d.png">
Here is an example. Note: the program being analyzed is for Cortex-A57, and it runs programs on a non-posix linux-based firmware.

**Comment by pjsoberoi** [comment]
Add this to the end of your AARCH64instructions.sinc:
`# FE DE FF E7
:und_bkpt
is b_0031=0xE7FFDEFE
{
}`
I'm not 100% sure on the endianess, you can swap it if it's wrong.

**Comment by astrelsky** [comment]
> @GameCubeGBA  At first glance, this seems to me like the correct behavior.  Bytes after a breakpoint (whether implemented as a HLT, BRK, UND, or INT) cannot generally be assumed to be valid opcodes as there is no presumption of flow across the breakpoint.  Admittedly, I am more familiar with x86 than arm64, but, in x86 code, breakpoint equivalents are regularly used to delineate code boundaries.  Disassembling through them would produce erroneous results.  Can you provide a specific example where there is intended flow across a breakpoint?

It appears to be very common for this instruction to remain in release builds of arm/aarch64 programs. When running the program the instruction either does nothing or has no effect on the behavior of the program. Leaving them as an undefined instruction (and therefore an error bookmark in the listing) breaks the flow of the disassembler and the decompiler.

**Comment by GhidorahRex** [comment]
We have this instruction in the base ARM 32 spec, but it is missing from the AARCH64. The instruction is listed in the manual, so the bytes should at least disassemble correctly.

**Comment by d-millar** [comment]
I guess this breaks down into two questions.  Per @pjsoberoi and @GhidorahRex, do we need to add this to the AARCH64 language def, and per @astrelsky is this flow-through (vs terminal).  Seems like the answer might be yes and yes.

**Comment by GhidorahRex** [comment]
@d-millar I can't find a reference to this instruction being used inAARCH64, so I'm not comfortable adding the instruction to the sleigh spec yet. But I'll keep digging into it.

**Comment by astrelsky** [comment]
> @d-millar I can't find a reference to this instruction being used inAARCH64, so I'm not comfortable adding the instruction to the sleigh spec yet. But I'll keep digging into it.

I'm no expert on arm/aarch64 so take this with a grain of salt. I think what this *really* is really doing is causing an undefined instruction hardware exception and then the handler in the interrupt vector handles it accordingly.

**Comment by d-millar** [comment]
@GhidorahRex yeah, the Debug section of the reference manual makes no obvious references to this - interesting read though.

**Comment by GhidorahRex** [comment]
> > @d-millar I can't find a reference to this instruction being used inAARCH64, so I'm not comfortable adding the instruction to the sleigh spec yet. But I'll keep digging into it.
> 
> I'm no expert on arm/aarch64 so take this with a grain of salt. I _think_ what this is really doing is causing an undefined instruction hardware exception and then the handler in the interrupt vector handles it accordingly.

I understand that part of the issue. The part I'm hesitant about is that I can find no reference to this instruction being associated with the AARCH64, only with ARM32. In ARM32, the instruction disassembles appropriately in Ghidra.

**Comment by d-millar** [comment]
So, https://usermanual.wiki/Document/ARMarmarchitecturereferencemanual.1897755950/view has UDF at A8.8.247 - still looking for UND

**Comment by GhidorahRex** [comment]
@d-millar That is ARM32 (and it is included in ARMinstructions.sinc on line 1620), not AARCH64

**Comment by d-millar** [comment]
Yeah, yeah - I hear ya (again, "still looking")

**Comment by d-millar** [comment]
Better ref for A64's UDF:  https://developer.arm.com/documentation/ddi0596/2021-09/Base-Instructions/UDF--Permanently-Undefined-?lang=en

**Comment by d-millar** [comment]
OK, well, have at least found the source for this (as noted by @GhidorahRex, not in any Arm Reference Manual).  This byte sequence is actually defined in the Linux source (ptrace.h) to create a PTRACE_BREAKPOINT.  No argument being made here for what action should be taken.

**Comment by GameCubeGBA** [comment]
I would suggest adding it as a breakpoint function. However, it is linux-specific, though if this comes up chances are it is a binary on a linux based OS anyway

**Comment by astrelsky** [comment]
> I would suggest adding it as a breakpoint function. However, it is linux-specific, though if this comes up chances are it is a binary on a linux based OS anyway

I have seen it on non linux systems

**Comment by GameCubeGBA** [comment]
So how do we go about this, if I may ask?

**Comment by pjsoberoi** [comment]
Obviously it's up to the Ghidra team on how to proceed. Assuming that this instruction is not an actual CPU instruction and is instead an undefined instruction that is caught by the debugger, I think the correct answer is to:

1) Still add the instruction to the slaspec as "und_bkpt". With und meaning undefined. 
2) Instead of having it act as a NOP as in my previous post, define it as "unimpl". This behavior will prevent further disassembly upon hitting the instruction. Just like as if the instruction was never defined in the first place. Someone wanting to disassemble after can manually make Ghidra disassemble further instructions. 

Ex: 
```
   # FE DE FF E7
   :und_bkpt
   is b_0031=0xE7FFDEFE
   unimpl
```

**Comment by GhidorahRex** [comment]
I still haven't seen any convincing evidence that points to this instruction being an AARCH64-bit trap instructions. All of the linux code I've seen referenced point to it being an ARM 32-bit undefined breakpoint, whereas the AARCH64 has a different one. I've spent a couple hours googling this and came up empty.

I am more 'descriptivist' than prescriptivist when it comes to documenting instruction format. I would like to see another sample from a different source, or some linux code that specifically references this instruction in ARM 64.

**Comment by ryanmkurtz** [comment]
The commit that closed this was a mistake.  Issue is still open/unresolved.

**Comment by M1CR0H4CK3R** [comment]
@GhidorahRex is there an ETA for this? I'm decompiling Nintendo Switch executables and hitting a roadblock with this instruction. I set it to dword data type to work around the erroneous bookmark being recreated upon deletion, but would really prefer it if the instruction were supported natively.

**Comment by pjsoberoi** [comment]
> @GhidorahRex is there an ETA for this? I'm decompiling Nintendo Switch executables and hitting a roadblock with this instruction. I set it to dword data type to work around the erroneous bookmark being recreated upon deletion, but would really prefer it if the instruction were supported natively.

What's wrong with this workaround: https://github.com/NationalSecurityAgency/ghidra/issues/3723#issuecomment-988481798

**Comment by GhidorahRex** [comment]
@pjsoberoi I believe that's what we're ultimately going to go with. This issue has been put off because the instruction is not defined in the aarch64 spec and no disassemblers that I can find for AARCH64 disassemble it as such. However, there does seem to be evidence in this thread that it's being used in aarch64 binaries as a breakpoint. I will push it up again.

**Comment by sad-dev** [comment]
This reminds me of a Q&A I put up just a while ago (https://github.com/NationalSecurityAgency/ghidra/discussions/6878)

Whether the bytes lead to halt/nop/interrupt handler/etc. might vary from binary to binary. If it is possible, I would like to be able to specify the Pcode of the instruction for a given binary without needing to touch the sinc/sla files

**Comment by pjsoberoi** [comment]
> @pjsoberoi I believe that's what we're ultimately going to go with. This issue has been put off because the instruction is not defined in the aarch64 spec and no disassemblers that I can find for AARCH64 disassemble it as such. However, there does seem to be evidence in this thread that it's being used in aarch64 binaries as a breakpoint. I will push it up again.

@GhidorahRex: My response was for [M1CR0H4CK3R](https://github.com/M1CR0H4CK3R) so he doesn't have to wait for an update. In general I lean towards what you were saying, if it's not a documented instruction I don't think it should be supported by default. My two cents. 

Maybe the Switch community can fork the processor module (assuming there are other Switch specific changes for the processor).
